### ISSUE TITLE
Form controls: labeling required fields

### ISSUE THREAD

**Issue by jAigret-Bix** [uswds/uswds#3691]
USAGov accessibility [test results ](https://docs.google.com/document/d/1g5beRCFgb5M7xYrEpBPuZy39AKYxtCXzT54FHTPiRgg/edit)

See associated bug issue #3788 
See associate issue https://github.com/uswds/uswds/issues/3295

1. Instructions of the form not placed at the top of the page (https://www.w3.org/TR/UNDERSTANDING-WCAG20/minimize-error-cues.html) . Place instructions for a form at the top of the page. *For example “All fields are required except optional”*
2. Also, required fields not identified. Users should be able to easily determine which data entry fields are required and which are optional. (https://www.w3.org/TR/2016/NOTE-WCAG20-TECHS-20161007/H90)

See Zendesk ticket #20401 w question: "will marking optional fields only pass the Trusted Tester process and 508 compliance? Also, if all fields are required, is it safe to leave that out of the instructions?"

[Slack](https://gsa-tts.slack.com/archives/C050HRGN7/p1601400724037300): The form required/optional one is a little bit more tricky, perhaps, since it'll also require updating our guidance. See https://github.com/uswds/uswds/issues/3295

- [x] @allie-shaw  to research and develop recommendations. 
- [x] meet with accessibility testers
- [x] review and capture recommendations 
- [ ] 
- [x] add instructions and templates
- [x] Allie - content audit - check and verify all form controls that we don't have conflicting guidance
- [ ] publish

**Comment by allie-shaw** [comment]
## Recommendations
Do not have one error message at the top (ie: All fields required)
- People don’t read instructions at the top of forms.
- Even if people read instructions, they may forget them. 
- People have to scan the form to determine if the field is required. 

Instead:
- Inputs should be clearly indicated as optional or required. These indicators can be included in the label. 
- If most fields in the form are required, indicate only the optional fields.
- If most fields in the form are optional, indicate only the required fields.

_______________________________________________________
### Carbon Design System
https://www.carbondesignsystem.com/components/form/usage#validation-and-errors
Optional versus required fields
All fields in a form are assumed required, with optional fields being tagged as so. It is recommended that an instruction precedes the form stating either “All fields are required” or “All fields are required unless marked optional”.

### Material Design System
https://material.io/components/text-fields
**Required text indicator**
To indicate that a field is required, display an asterisk (*) next to the label text and mention near the form that asterisks indicate required fields.
- If some fields are required, indicate all required ones
- If most fields are required, indicate optional fields by displaying the word “optional” in parentheses next to the label text
- If required text is colored, that color should also be used for the asterisk

### NNG
https://www.nngroup.com/articles/required-fields/
Usually designers will do one of the following:
- They show instructions at the top of the form saying All fields are required or All fields are required unless otherwise indicated.
- They mark the optional fields, since they are usually fewer.

**Issues with this approach**
People don’t read instructions at the top of forms.
Even if people read instructions, they may forget them. 
People have to scan the form to determine if the field is required. 

The solution is simple: mark all the required fields. Be as explicit and transparent as possible: for every single field that must be completed, mark it as required.

### Canadian Government Design System
https://design.gccollab.ca/component/overall-design-and-labels
**Required Fields**
Inputs should be clearly indicated as optional or required. These indicators can be included in the label. 
If most fields in the form are required, indicate only the optional fields.
If most fields in the form are optional, indicate only the required fields.

### Singapore Government Design System
https://www.designsystem.tech.gov.sg/docs/general-guidelines/

- Label fields as optional instead of compulsory 
- Avoid using asterisks to mark compulsory fields

### The Anatomy of Accessible Forms: Required Form Fields
https://www.deque.com/blog/anatomy-of-accessible-forms-required-form-fields/

- How to indicate a required field
- Provide the required text in the label.
- Provide a graphic * image in the label with appropriate alt text.
- Providing a star (asterisk) symbol.
- Use of color to identify if a form control is required.
- Providing HTML5 and ARIA required attributes.

### ARIA Labels and Relationships
https://developers.google.com/web/fundamentals/accessibility/semantics-aria/aria-labels-and-relationships

aria-labelledby
aria-labelledby allows us to specify the ID of another element in the DOM as an element's label.
 
This is much like using a label element, with some key differences.
1.	aria-labelledby may be used on any element, not just labelable elements.
2.	While a label element refers to the thing it labels, the relationship is reversed in the the case of aria-labelledby — the thing being labeled refers to the thing that labels it.
3.	Only one label element may be associated with a labelable element, but aria-labelledby can take a list of IDREFs to compose a label from multiple elements. The label will be concatenated in the order that the IDREFs are given.
4.	You can use aria-labelledby to refer to elements that are hidden and would otherwise not be in the accessibility tree. For example, you could add a hidden span next to an element you want to label, and refer to that with aria-labelledby.
5.	However, since ARIA only affects the accessibility tree, aria-labelledby does not give you the familiar label-clicking behavior you get from using a label element.
Importantly, aria-labelledby overrides all other name sources for an element. So, for example, if an element has both an aria-labelledby and an aria-label, or an aria-labelledby and a native HTML label, the aria-labelledby label always takes precedence.

**Comment by jAigret-Bix** [comment]
Trusted Tester meting 10/29. Recommendations have been captured in above comment @allie-shaw

**Comment by mejiaj** [comment]
Voiceover _is_ ignoring both `aria-labelledby` and `aria-describedby`. I was able to get it to read the alert message by removing `role` attribute from alert.

Created a test available [here →](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.app.cloud.gov/preview/uswds/uswds/jm-form-alerts-voiceover/components/preview/input-alerts.html)

**Comment by jaredcunha** [comment]
We are working on some templates for signing in and creating an account. Will this guidance apply to smaller forms with 2-3 fields, and the required fields might be a little more implicit? For example, here's a sign in form. 

<img width="525" alt="Screen Shot 2020-11-16 at 3 45 36 PM" src="https://user-images.githubusercontent.com/1094951/99310549-2e455d00-2829-11eb-9192-74d39ac5efd6.png">

**Comment by allie-shaw** [comment]
Hi Jared,

Thanks for sharing. I have seen similar for smaller forms (such as log in screens) and not having the required forms noted. However, we checked with our Trusted Tester and it is still recommended to indicated the required fields. WCAG does not provide exceptions for forms with only [X] number of forms or anything like that.

I had also checked about using placeholder text instead but it is not recommened for several reasons, including the fact that not all AT supports it and the grey text usually fails contrast standards.

So, in short, we still recommend indicating the required fields (either with “required” or “*”).

Allie

**Comment by jaredcunha** [comment]
@allie-shaw Thanks! That's really helpful. I can update our templates to reflect this. You mentioned above that describing the required fields should be done both at the top and in-line with the label. Is the text okay as I have in the screenshot? Or should is there some other way I should word it (there won't be any options fields in these templates).

**Comment by allie-shaw** [comment]
Hi Jared,

What you have in the mockup is fine. I’ve also seen examples where it’s a shorter description at the top.
IE: “* indicates a required field” or even just “* required information”

I hope that helps.

Allie

**Comment by thisisdano** [comment]
I will say that it seems weird, redundant, and potentially silly to do this on a signin form like what we see above. We can move forward with it, but let's not get ourselves in the unusual situation of adding, say, a required link to a search box or a similar one-field form. I believe there is a reasonable practical limit to this guidance, a blanket requirement feels half-baked and inappropriate, despite what WCAG may say. The design systems and examples listed above tend to give a more nuanced approach, which I support.

**Comment by jAigret-Bix** [comment]
1/11: target for Feb monthly call. Code is ready, Guidance needs a closer look. create new issue for follow up on guidance change. target comms package for outreach strat.
